typedData/typedData.go

```diff
@@ -0,0 +1,896 @@
+		strValue := fmt.Sprintf("%v", data)
+		bigNum, ok := new(big.Int).SetString(strValue, 0)
+		if !ok {
+			return resp, fmt.Errorf("faild to convert '%s' of type 'i128' to big.Int", strValue)
+		}
+		feltValue := new(felt.Felt).SetBigInt(bigNum)
+		return feltValue, nil
+	case "string":
+		if rev.Version() == 0 {
+			resp, err := getFeltFromData()
+			if err != nil {
+				return resp, err
+			}
+			return resp, nil
+		} else {
+			value := fmt.Sprintf("%v", data)
+			byteArr, err := utils.StringToByteArrFelt(value)
+			if err != nil {
+				return resp, err
+			}
+			return rev.HashMethod(byteArr...), nil
+		}
+	case "selector":
+		value := fmt.Sprintf("%v", data)
+		return utils.GetSelectorFromNameFelt(value), nil
+	default:
+		return resp, fmt.Errorf("invalid type '%s'", typeName)
+	}
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface for TypedData
+func (td *TypedData) UnmarshalJSON(data []byte) error {
+	var dec map[string]json.RawMessage
+	if err := json.Unmarshal(data, &dec); err != nil {
+		return err
+	}
+
+	// primaryType
+	primaryType, err := utils.GetAndUnmarshalJSONFromMap[string](dec, "primaryType")
+	if err != nil {
+		return err
+	}
+
+	// domain
+	domain, err := utils.GetAndUnmarshalJSONFromMap[Domain](dec, "domain")
+	if err != nil {
+		return err
+	}
+
+	// types
+	rawTypes, err := utils.GetAndUnmarshalJSONFromMap[map[string]json.RawMessage](dec, "types")
+	if err != nil {
+		return err
+	}
+	var types []TypeDefinition
+	for key, value := range rawTypes {
+		var params []TypeParameter
+		if err := json.Unmarshal(value, &params); err != nil {
+			return err
+		}
+
+		typeDef := TypeDefinition{
+			Name:       key,
+			Parameters: params,
+		}
+
+		types = append(types, typeDef)
+	}
+
+	// message
+	rawMessage, ok := dec["message"]
+	if !ok {
+		return fmt.Errorf("invalid typedData json: missing field 'message'")
+	}
+	bytesMessage, err := json.Marshal(rawMessage)
+	if err != nil {
+		return err
+	}
+
+	// result
+	resultTypedData, err := NewTypedData(types, primaryType, domain, bytesMessage)
+	if err != nil {
+		return err
+	}
+
+	*td = *resultTypedData
+	return nil
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface for Domain
+func (domain *Domain) UnmarshalJSON(data []byte) error {
+	var dec map[string]any
+	if err := json.Unmarshal(data, &dec); err != nil {
+		return err
+	}
+
+	getField := func(fieldName string) (string, error) {
+		value, ok := dec[fieldName]
+		if !ok {
+			return "", fmt.Errorf("error getting value of '%s' from 'domain' struct", fieldName)
+		}
+		return fmt.Sprintf("%v", value), nil
+	}
+
+	name, err := getField("name")
+	if err != nil {
+		return err
+	}
+
+	version, err := getField("version")
+	if err != nil {
+		return err
+	}
+
+	chainId, err := getField("chainId")
+	if err != nil {
+		var err2 error
+		// ref: https://community.starknet.io/t/signing-transactions-and-off-chain-messages/66
+		chainId, err2 = getField("chain_id")
+		if err2 != nil {
+			return err
+		}
+	}
```

Comment:
Maybe something like

value, err1 := getField("chainId")
    if err1 == nil {
        return value, nil
    }

    value, err2 := getField("chain_id")
    if err2 == nil {
        return value, nil
    }

    // Return both errors if both fail
    combinedError := fmt.Errorf("primary error: %v; fallback error: %v", err1, err2)
    return "", combinedError

Reply:
This was intentional.
The SNIP says the correct is `chainId` but I [found ](https://github.com/starknet-io/SNIPs/issues/122)that previously the `chain_id` name was also used. So it will first look for the 'chainId' field, if not found, it'll then look for the 'chain_id'. If neither 'chainId' or 'chain_id' were found, the error will only say about the `chainId` field since it's the currently used field, the 'chain_id' is verified only for backward compatibility. WDYT?

Now that you have mentioned, I believe it'll be good to before looking for the 'chain_id' field it'll first verify if the Revision version is 0 since the only acceptable field for Revision 1 is 'chainId', so it would be invalid in this case.

Reply:
Yup, I think we should check the revision, and if it's 0 then return both errors, and if it's 1 just check if it's `chainId` (assuming revision 1 only allows chainId)

Reply:
Done